[18 India]  Implement GIPR Differences and Closing a Corporation

EdPovi opened the pull request:
- [x] Branch is derived from the latest `master`
- n/a Add the `pins` or `archive_alpha_games` label if this change will break existing games
- [x] Code passes linter with `docker compose exec rack rubocop -a`
- [x] Tests pass cleanly with `docker compose exec rack rake`

## Implementation Notes
**GIPR Differences:**
- Can’t Operate until phase II 
- Place home token in any home city or
- Place home token in in upgraded town -> green single city
- Use exchange tokens when another Corp closes

**Closing a Corporation:**
- Trains returned to market
- Corp owned shares / privates / bonds returned to market
- Remove all shares of self
- Remove all proxy companies from hand / IPO rows
- Return treasury to bank
- Adjust cert limit

Adjusted commodity concession methods to allow for display on company card. 
- Implemented by modifying the description of an Ability

Train Tweaks:
- May not own more than one Phase IV train
- May not sell the 4(E) x 3 train

OO Yellow Tile fix:
- When the yellow OO tile is built, the laying corp may select the city for existing token.
- BUT the base code doesn't account for possibility that the other city is reserved rather than tokened. 

### Any Assumptions / Hacks
Since using exchange tokens is a choice is a for non-home tokens of closing company, created an "ExchangeToken" step based off of HomeToken and creating a "pending_exchange_tokens" round state to trigger the step.

HomeTrack is a modified HomeToken step to allow for town tiles to be upgraded to a green single city prior to placing home token.

ollybh commented on lib/engine/game/g_18_india/game.rb:
```diff
@@ -790,49 +988,51 @@ def commodity_bonus(route, _stops)
           commodity_sources.each do |source|
             bonus = 0
+            @round.commodities_used = []
             case source
```
There's a lot of repetitive code in this section, and hardcoded values. Could this be first bundling together all the information in one place:
```
COMMODITY_BONUSES = {
  'OIL' => { value: 30, commodity: 'OIL', locations: ['MUMBAI'] },
  'OPIUM' => { value: 100, commodity: 'OPIUM', locations: %w[LAHORE HALDIA] },
  'ORE1' => { value: 50, commodity: 'ORE', locations: %w[KARACHI CHENNAI] },
  # ...
}.freeze
```
And then, in `commodity_bonus`:
```
commodity_sources.each do |source|
  bonus = COMMODITY_BONUSES[source]
  if visited_names.intersect?(bonus[locations])
    revenue += bonus[value]
    @round.commoditites_used << bonus[commodity]
  end
end

EdPovi replied:
I think I understand your suggestion, I'll see if I can refactor to that approach.

EdPovi replied:
I have refactored to this method.  It is cleaner and seems to work well.

ollybh reviewed (changes requested):
This mostly looks good, but there are a few things that could be improved.
Throughout out this you have several places where you are testing `@round.phase` for a certain value. Have you considered using phase statuses for things like this? It would mean you aren't hardcoding the phase names in the code, and it would also make the timing visible to players on the game info page.

EdPovi commented on lib/engine/game/g_18_india/game.rb:
```diff
@@ -630,21 +649,62 @@ def remove_from_hand(player, company)
         end
 
+        def open_city_hexes
+          hexes = @cities.reject { |c| c.available_slots.zero? }.map { |c| c.tile.hex }.compact
+          hexes.uniq
```
I recall testing some results via the debug LOGGER, and seeing nills in the array.  But I will test to confirm if this works without the compact call.

EdPovi replied:
If I remove compact, a nil result is present:
![image](https://github.com/tobymao/18xx/assets/118568589/4e1c5ddb-1a38-4e0e-ae84-c1fe954aa77d)

EdPovi replied:
I have reformatted the line as you suggested.

EdPovi commented on lib/engine/game/g_18_india/game.rb:
```diff
@@ -656,13 +716,144 @@ def float_corporation(corporation)
+              @bank.companies.push(company)
+            end
+            corporation.companies.delete(company)
+          end
```
That should work, I'll test it.

EdPovi commented on lib/engine/game/g_18_india/game.rb:
```diff
@@ -656,13 +716,144 @@ def float_corporation(corporation)
+        end
+
+        def gipr_exchange_with_closing_corp(corporation)
+          corporation.tokens.dup.each_with_index do |token, index|
+            LOGGER.debug "gipr_exchange > #{corporation.tokens} && index-#{index} used-#{token.used}"
+            break unless token.used
```
I'll see if I can use placed_tokens, as that should remove the need for the `used` guard clause.   I think I needed the `dup` because I was deleting the tokens as I went through the list.   I recall this flow being a bit complex as `close_corporation` would get called again with the way pending tokens worked, so I needed a way to handle re-entering that method to check if we were already did the exchange token test.

EdPovi replied:
Using` corporation.placed_tokens` works.  I also removed the `dup` call and changed the token `destroy` to `remove`, which just modifies the token.  Seems to work find in testing.

EdPovi commented on lib/engine/game/g_18_india/game.rb:
```diff
@@ -630,21 +649,62 @@ def remove_from_hand(player, company)
+        end
+
+        def town_to_green_city_hexes
+          @hexes.map(&:tile).filter do |t|
```
I changed it to select, as filter is an alias.

EdPovi commented:
Changed some of the `phase.name` checks to `phase.status.include?`.  

I believe I have addressed your comments and suggestions.  And should be ready for your review.

Thanks!

ollybh reviewed (approved):
Thanks for making these changes.
